lib/Sema/OpenedExistentials.cpp

```diff
@@ -644,14 +637,11 @@ swift::canOpenExistentialCallArgument(ValueDecl *callee, unsigned paramIdx,
   if (genericParam->getDepth() < genericSig->getMaxDepth())
     return std::nullopt;
 
-  Type existentialTy;
-  if (auto existentialMetaTy = argTy->getAs<ExistentialMetatypeType>())
-    existentialTy = existentialMetaTy->getInstanceType();
-  else
-    existentialTy = argTy;
-
-  ASSERT(existentialTy->isAnyExistentialType());
-
+  // The binding could be an existential metatype. Get the instance type for
+  // conformance checks and to build an opened existential signature. If the
+  // instance type is not an existential type, i.e., the metatype is nested,
+  // bail out.
+  const Type existentialTy = bindingTy->getMetatypeInstanceType();
```

Comment:
Looking through a metatype here causes us to open the existential in case 1, but not case 2:

```swift
func foo<T : BitwiseCopyable>(_: T) -> T {}

// 1
do {
  let exist: any Any.Type
  let result = foo(exist)
}

// 2
do {
  let exist: any Any.Type.Type
  let result = foo(exist)
}
```

I took care to not change the behaviour, but it clearly contradicts [SE-0352](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0352-implicit-open-existentials.md#avoid-opening-when-the-existential-type-satisfies-requirements-in-swift-5). I suppose this became a problem when metatypes started conforming to these marker protocols. @DougGregor Do you think we could risk correcting this?

Btw, the standard library [seems to be making use](https://github.com/swiftlang/swift/blob/f1f10517d5b6f4e07b316f53fb50f0f8b2928b6a/stdlib/public/core/KeyPath.swift#L3769) of the current case 1 behaviour.